File: internal/screener/screener.go
package screener

import (
	"context"
	"fmt"
	"math"
	"time"

	"github.com/Ruscigno/stock-screener/internal/config"
	"github.com/Ruscigno/stock-screener/internal/extrema"
	"github.com/Ruscigno/stock-screener/internal/indicators"
	"github.com/Ruscigno/stock-screener/internal/match"
	"github.com/Ruscigno/stock-screener/internal/resample"
	"github.com/Ruscigno/stock-screener/internal/storage"
	"github.com/Ruscigno/stock-screener/internal/timeframe"
)

// trendEpsilon is the dead-band for the rising/falling/flat slope. It is
// intentionally near-zero: for continuous indicator values an exact tie between
// two bars is vanishingly rare, so trend is effectively always rising/falling —
// which is the directional signal we want. "flat" is reserved for genuine
// equality (e.g. constant/insufficient series). If a wider dead-band is ever
// needed, promote this to a configurable screening setting.
const trendEpsilon = 1e-9

type Screener struct {
	store storage.Store
	cfg   *config.Config
}

func New(store storage.Store, cfg *config.Config) *Screener {
	return &Screener{store: store, cfg: cfg}
}

func (s *Screener) Screen(ctx context.Context, req Request) (Result, error) {
	var res Result
	for _, symbol := range req.Symbols {
		for _, tfName := range req.Timeframes {
			tf, ok := timeframe.Get(tfName)
			if !ok {
				res.Warnings = append(res.Warnings, Warning{symbol, tfName, "unknown timeframe"})
				continue
			}
			bars, err := s.loadBars(ctx, symbol, tf)
			if err != nil {
				res.Warnings = append(res.Warnings, Warning{symbol, tfName, "load error: " + err.Error()})
				continue
			}
			if len(bars) == 0 {
				res.Warnings = append(res.Warnings, Warning{symbol, tfName, "no_data"})
				continue
			}
			row, warns := s.evaluate(symbol, tfName, bars, req)
			res.Warnings = append(res.Warnings, warns...)
			if row != nil {
				res.Rows = append(res.Rows, *row)
			}
		}
	}
	return res, nil
}

func (s *Screener) loadBars(ctx context.Context, symbol string, tf timeframe.TF) ([]storage.Bar, error) {
	need := requiredBars(s.cfg, tf)
	if tf.Native {
		return s.store.GetBars(ctx, symbol, tf.Name, need)
	}
	// Derived TFs aggregate GroupSize parent bars each, so fetch proportionally
	// more parent bars before resampling.
	parent, err := s.store.GetBars(ctx, symbol, tf.Parent, need*tf.GroupSize)
	if err != nil {
		return nil, err
	}
	if s.cfg.Collector.UseClosedBarsOnly {
		return resample.ToClosed(parent, tf.Name), nil
	}
	return resample.To(parent, tf.Name), nil
}

// requiredBars is how many bars (in tf's own resolution) the screener loads per
// (symbol, timeframe). It is the larger of:
//   - the warmup the longest indicator needs plus room to confirm several
//     pivots (a hard floor, so indicators never come back all-NaN), and
//   - the peak_lookback window converted to bars (the configured minimum
//     history to scan for recent pivots).
//
// Bounding the load this way keeps each /screen request O(window) instead of
// O(full stored history).
func requiredBars(cfg *config.Config, tf timeframe.TF) int {
	longest := cfg.Indicators.RSI.Length
	if l := cfg.Indicators.VolumeOscillator.LongLength; l > longest {
		longest = l
	}
	if l := cfg.Indicators.DistanceFromMA.Length; l > longest {
		longest = l
	}
	warmup := longest + (2*cfg.Screening.PivotWindow+1)*(cfg.Screening.PeaksToShow+1) + 50

	lookbackBars := 0
	if tf.BarDuration > 0 {
		lookbackBars = int(time.Duration(cfg.Screening.PeakLookback) / tf.BarDuration)
	}
	if lookbackBars > warmup {
		return lookbackBars
	}
	return warmup
}

func (s *Screener) evaluate(symbol, tfName string, bars []storage.Bar, req Request) (*Row, []Warning) {
	var warns []Warning
	closes := make([]float64, len(bars))
	volumes := make([]float64, len(bars))
	for i, b := range bars {
		closes[i] = b.Close
		volumes[i] = b.Volume
	}
	w := s.cfg.Screening.PivotWindow
	nShow := s.cfg.Screening.PeaksToShow
	results := map[string]IndicatorResult{}
	var triggered []string

	for _, ind := range req.Indicators {
		series := s.series(ind, closes, volumes)
		if series == nil {
			warns = append(warns, Warning{symbol, tfName, "unknown indicator: " + ind})
			continue
		}
		idx := lastNonNaNScreener(series)
		if idx < 0 {
			warns = append(warns, Warning{symbol, tfName,
				fmt.Sprintf("insufficient_data: %s has no value (need more bars)", ind)})
			continue
		}
		peaks := extrema.LastN(extrema.FindPeaks(series, w), nShow)
		valleys := extrema.LastN(extrema.FindValleys(series, w), nShow)
		zone := classify(series[idx], peaks, valleys)
		ir := IndicatorResult{
			Latest:    series[idx],
			Trend:     trend(series, idx, s.cfg.Screening.TrendLookback),
			Zone:      zone,
			Triggered: zone != "neutral",
			Peaks:     toPoints(peaks, bars),
			Valleys:   toPoints(valleys, bars),
		}
		results[ind] = ir
		if ir.Triggered {
			triggered = append(triggered, ind)
		}
	}

	if !match.Qualifies(len(triggered), len(req.Indicators), req.Match) {
		return nil, warns
	}
	last := bars[len(bars)-1]
	return &Row{
		Symbol: symbol, Timeframe: tfName, BarTime: last.Time, Price: last.Close,
		Triggered: triggered, Indicators: results,
	}, warns
}

func (s *Screener) series(ind string, closes, volumes []float64) []float64 {
	switch ind {
	case IndRSI:
		return indicators.RSI(closes, s.cfg.Indicators.RSI.Length)
	case IndVolOsc:
		return indicators.VolumeOscillator(volumes,
			s.cfg.Indicators.VolumeOscillator.ShortLength,
			s.cfg.Indicators.VolumeOscillator.LongLength)
	case IndDistance:
		return indicators.DistanceFromMA(closes,
			s.cfg.Indicators.DistanceFromMA.MAType,
			s.cfg.Indicators.DistanceFromMA.Length)
	}
	return nil
}

func trend(series []float64, idx, lookback int) string {
	prev := idx - lookback
	if prev < 0 || math.IsNaN(series[prev]) {
		return "flat"
	}
	diff := series[idx] - series[prev]
	switch {
	case diff > trendEpsilon:
		return "rising"
	case diff < -trendEpsilon:
		return "falling"
	default:
		return "flat"
	}
}

func toPoints(pivots []extrema.Pivot, bars []storage.Bar) []PivotPoint {
	out := make([]PivotPoint, 0, len(pivots))
	for _, p := range pivots {
		out = append(out, PivotPoint{Value: p.Value, Time: bars[p.Index].Time})
	}
	return out
}

func lastNonNaNScreener(v []float64) int {
	for i := len(v) - 1; i >= 0; i-- {
		if !math.IsNaN(v[i]) {
			return i
		}
	}
	return -1
}

File: stock_screener/screener.py
from __future__ import annotations
from dataclasses import dataclass, field
from datetime import datetime
import pandas as pd
from stock_screener import timeframes, indicators, detection, resample
from stock_screener.config import Config, parse_duration
from stock_screener.rule import classify, qualifies

ALL_INDICATORS = ["rsi", "volume_oscillator", "distance_from_ma"]

@dataclass
class Pivot:
    value: float
    time: datetime

@dataclass
class IndicatorResult:
    latest: float
    trend: str
    zone: str
    triggered: bool
    peaks: list[Pivot]
    valleys: list[Pivot]

@dataclass
class Row:
    symbol: str
    timeframe: str
    bar_time: datetime
    price: float
    triggered: list[str]
    indicators: dict[str, IndicatorResult]

@dataclass
class Warning:
    symbol: str
    timeframe: str
    message: str

@dataclass
class Result:
    rows: list[Row] = field(default_factory=list)
    warnings: list[Warning] = field(default_factory=list)

class Screener:
    def __init__(self, store, cfg: Config):
        self.store = store
        self.cfg = cfg

    def screen(self, symbols, timeframes_, match, indicators_) -> Result:
        res = Result()
        for symbol in symbols:
            for tf_name in timeframes_:
                tf = timeframes.get(tf_name)
                if tf is None:
                    res.warnings.append(Warning(symbol, tf_name, "unknown timeframe"))
                    continue
                df = self._load(symbol, tf)
                if df is None or df.empty:
                    res.warnings.append(Warning(symbol, tf_name, "no_data"))
                    continue
                row, warns = self._evaluate(symbol, tf_name, df, match, indicators_)
                res.warnings.extend(warns)
                if row is not None:
                    res.rows.append(row)
        return res

    def _load(self, symbol, tf):
        need = self._required_bars(tf)
        if tf.native:
            return self.store.get_bars(symbol, tf.name, need)
        parent = self.store.get_bars(symbol, tf.parent, need * tf.group_size)
        if parent is None or parent.empty:
            return parent
        if self.cfg.collector.use_closed_bars_only:
            return resample.to_closed(parent, tf.name)
        return resample.to(parent, tf.name)

    def _required_bars(self, tf) -> int:
        longest = max(self.cfg.indicators.rsi.length,
                      self.cfg.indicators.volume_oscillator.long_length,
                      self.cfg.indicators.distance_from_ma.length)
        warmup = longest + 5 * (self.cfg.screening.peaks_to_show + 1) + 50
        lookback_bars = parse_duration(self.cfg.screening.peak_lookback) // max(tf.bar_seconds, 1)
        return max(warmup, int(lookback_bars))

    def _series(self, ind, df):
        if ind == "rsi":
            return indicators.rsi(df["close"], self.cfg.indicators.rsi.length)
        if ind == "volume_oscillator":
            v = self.cfg.indicators.volume_oscillator
            return indicators.volume_oscillator(df["volume"], v.short_length, v.long_length)
        if ind == "distance_from_ma":
            d = self.cfg.indicators.distance_from_ma
            return indicators.distance_from_ma(df["close"], d.ma_type, d.length)
        return None

    def _min_bars(self, ind) -> int:
        if ind == "rsi":
            return self.cfg.indicators.rsi.length + 1
        if ind == "volume_oscillator":
            return self.cfg.indicators.volume_oscillator.long_length
        if ind == "distance_from_ma":
            return self.cfg.indicators.distance_from_ma.length
        return 0

    def _evaluate(self, symbol, tf_name, df, match, indicators_):
        warns: list[Warning] = []
        results: dict[str, IndicatorResult] = {}
        triggered: list[str] = []
        times = list(df.index)
        for ind in indicators_:
            raw = self._series(ind, df)
            if raw is None:
                warns.append(Warning(symbol, tf_name, f"unknown indicator: {ind}"))
                continue
            det = getattr(self.cfg.indicators, ind).detection
            series = detection.smooth(raw.reset_index(drop=True), det.smoothing)
            valid = series.dropna()
            if valid.empty:
                warns.append(Warning(symbol, tf_name,
                    f"insufficient_data: {ind} needs {self._min_bars(ind)} bars, have {len(df)}"))
                continue
            idx = int(valid.index[-1])
            peaks, valleys = detection.find_extrema(series, det.min_prominence, det.min_distance)
            peaks = detection.last_n(peaks, self.cfg.screening.peaks_to_show)
            valleys = detection.last_n(valleys, self.cfg.screening.peaks_to_show)
            zone = classify(float(series.iloc[idx]), peaks, valleys)
            ir = IndicatorResult(
                latest=float(series.iloc[idx]),
                trend=self._trend(series, idx),
                zone=zone,
                triggered=(zone != "neutral"),
                peaks=[Pivot(v, times[i]) for i, v in peaks],
                valleys=[Pivot(v, times[i]) for i, v in valleys],
            )
            results[ind] = ir
            if ir.triggered:
                triggered.append(ind)
        if not qualifies(len(triggered), len(indicators_), match):
            return None, warns
        return (Row(symbol, tf_name, times[-1], float(df["close"].iloc[-1]),
                    triggered, results), warns)

    def _trend(self, series, idx) -> str:
        prev = idx - self.cfg.screening.trend_lookback
        eps = self.cfg.screening.trend_flat_epsilon
        if prev < 0 or pd.isna(series.iloc[prev]):
            return "flat"
        diff = float(series.iloc[idx]) - float(series.iloc[prev])
        if diff > eps:
            return "rising"
        if diff < -eps:
            return "falling"
        return "flat"

def aggregate_matches(res: Result, symbols, timeframes_) -> list[dict]:
    by_sym: dict[str, dict] = {}
    for row in res.rows:
        a = by_sym.setdefault(row.symbol, {"tfs": set(), "inds": set()})
        a["tfs"].add(row.timeframe)
        a["inds"].update(row.triggered)
    out = []
    for sym in symbols:
        a = by_sym.get(sym)
        if not a:
            continue
        tfs = [t for t in timeframes_ if t in a["tfs"]]
        inds = [i for i in ALL_INDICATORS if i in a["inds"]]
        out.append({"symbol": sym, "timeframes": tfs, "indicators": inds})
    return out
